Applies all patches from dotnet/source-build for 5.0

dominoFire opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/384
Regression: No
* Last working version:   
* How are we preventing it in future: working more closely with source-build team

## Fix

Details: Applies all patches located at: https://github.com/dotnet/source-build/tree/master/patches/nuget-client

**NOTE**: This PR blocks https://github.com/dotnet/source-build/pull/1649/

### Details on patches

- ~0001-Set-final-5.4-branding-3100.patch~
Adds `IsEscrow = true` property when in source-build mode. No effect.

- 0002-experimental-nuget-patch.patch
Remove IL merged assemblies that conflicts with source-build. Seems related to https://github.com/dotnet/source-build/issues/739

- ~0003-MSBuild-package-versions-for-dotnet-Source-Build-285.patch~
Adds RepoApi (source-build package injection) to NuGet.Client repo. It is already integrated in this repo

- 0004-Pin-newtonsoft-json-to-a-specific-version.patch
Sets Newtonsoft.Json to 9.0.1; already done in NuGet.Client via defaults and RepoApi

- ~0005-New-version-of-ref-assemblies.patch~ 
Updates netfx reference assemblies to 1.0.0-preview.1; removed ref. assembly since 5.0 is only building netcore assemblies.

- 0006-Do-not-build-NET-Fx-binaries-in-source-build.patch
Avoid adding any mention to netframework TFMs to avoid conflicts with source-build

- 0007-Do-not-set-RID-in-source-build.patch
When compiling NuGet.Build.Tasks.Pack in source-build, avoid setting the RuntimeIdentifier to win7-x64. This RID is needed to copy runtime-specifc assets from System.Security.Cryptography.ProtectedData 4.3.0 to the final netcore NuGet.Build.Task.Pack package.

- ~0008-Remove-unneeded-feeds.patch~
Remove feeds that conflicts with source-build at those could introduce unneeded prebuilt packages. Those feeds are VS-specific and added conditionally with the help of `RestoreAdditionalProjectSources`

- 0009-Clear-VsToolsPath-for-source-build.patch
It seems that VSToolsPath confuses source-build, so it is set to blank 

- 0010-Remove-trailing-separator-for-PackagePath.patch
Removes trailing backslashes in project file. Known issue documented here: https://github.com/NuGet/Home/issues/9705

## Testing

Tests Added: No  
Reason for not adding tests: Build infrastructure, it only needs to have a full green build to validate changes
Validation: source-build [CI run](https://dev.azure.com/dnceng/public/_build/results?buildId=723305&view=logs&jobId=dd0961f5-eb75-5263-969b-8fed4b0393f9&j=e242d376-9fed-565d-2a2a-c4e65b5ab60e): 
All but MacOS are green builds. CI Mac pipeline failure seems to be related with dotnet/installer repo

Commit:
Added 3 patches

Commit:
Modified to preview1

Commit:
no rid in NuGet.Build.task.Pack when source-build'ing

Commit:
Appplied 4 patches

Commit:
restored xplat signing patch

Commit:
Removed backslashes in NuGet.Build.Tasks.Console

Commit:
Removed VS feeds from NuGet.Config

Commit:
Remove CodeAnalyzers on source-build

zivkan commented on build/bootstrap.proj:
```diff
@@ -7,8 +7,20 @@
+        </AdditionalFeed>
+        <AdditionalFeed Include="VsSdk">
+            <Url>https://pkgs.dev.azure.com/azure-public/vside/_packaging/vssdk/nuget/v3/index.json</Url>
+        </AdditionalFeed>
```
I'm not a fan of this.  Are we under an obligation to get source build patches down to zero? Can we require them to keep these feeds?  Why are these AzDO feeds a problem for them, but not the myget feeds?

dominoFire replied:
I've got [some green builds](https://dev.azure.com/dnceng/public/_build/results?buildId=724671&view=logs&j=5848992b-caa4-56fb-db92-b29e38357b04) on source-build CI with the NuGet feeds untouched. I think I can work with source-build team to discard this patch.

dominoFire replied:
@crummel Can we get rid of the patch that removes nuget-client feeds?

dominoFire replied:
Resolved as per this comment https://github.com/NuGet/NuGet.Client/pull/3436#discussion_r438985442

zivkan commented on build/common.props:
```diff
@@ -6,7 +6,7 @@
   <!-- When building on non-Windows machines, we need the reference assemblies to build full-framework assemblies. -->
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' AND '$(DotNetBuildFromSource)' == 'true'">
-     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0-alpha-5" PrivateAssets="All" />
+     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0-preview.1" PrivateAssets="All" />
```
1. We should just use the non-preview version: https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies/1.0.0

2. given the patch `0006-Do-not-build-NET-Fx-binaries-in-source-build.patch`, why does this matter?

dominoFire replied:
I think it was flagged as a prebuilt dependency in the past. I will upgrade to 1.0.0

dominoFire replied:
Updated to 1.0.0 with no problems

zivkan commented on src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/NuGet.VisualStudio.OnlineEnvironment.Client.csproj:
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <MinimumVisualStudioVersion>16.0</MinimumVisualStudioVersion>
     <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+    <VSToolsPath Condition="'$(DotNetBuildFromSource)' == 'true'"></VSToolsPath>
```
this project shouldn't even be built in source build/on linux. Looks like we need to improve our build scripts.

dominoFire replied:
source-build builds XPLAT projects:

https://github.com/dotnet/source-build/blob/1eba1a673a160ae4635270fdcf823da4bfe66d3d/repos/nuget-client.proj#L46

See [logs](https://dev.azure.com/dnceng/public/_build/results?buildId=723305&view=artifacts&type=publishedArtifacts) at ubuntu1804/source-build/logs/artifacts/logs/nuget-client.log 

It seems that the variable is overridden when MSBuild evaluates which projects are XPlat and somewhere else, this variable is not set correctly.

dominoFire replied:
Actual error without the patch (edited for readability):

```
/src/artifacts/src/NC/src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/NuGet.VisualStudio.OnlineEnvironment.Client.csproj(101,3): error MSB4019: The imported project "/src/.dotnet/sdk/5.0.100-preview.4.20258.7/Microsoft/VisualStudio/v16.0/VSSDK/Microsoft.VsSDK.targets" was not found. Confirm that the expression in the Import declaration "/src/.dotnet/sdk/5.0.100-preview.4.20258.7/Microsoft/VisualStudio/v16.0/VSSDK/Microsoft.VsSDK.targets" is correct, and that the file exists on disk. [/src/artifacts/src/NC/src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/NuGet.VisualStudio.OnlineEnvironment.Client.csproj]
```

The import below tries to import VsSDK.targets file, which it does not exists in .NET Core. I'm not sure how can avoid evaluating the whole project file and only get the properties we need. Then, it seems that this patch is necessary.

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj:
```diff
@@ -10,7 +10,7 @@
     <PackProject>true</PackProject>
     <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
-    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
+    <RuntimeIdentifier Condition="'$(DotNetBuildFromSource)' != 'true'">win10-x64</RuntimeIdentifier>
```
@nkolev92 any ideas why we set runtime for our pack tasks package? I'm a little surprised it doesn't cause problems on mac/linux.

nkolev92 replied:
no clue tbh.

dominoFire replied:
The first commit with this file with a RuntimeIdentifier refers to a commit to ilmerge 
https://github.com/NuGet/NuGet.Client/commit/38949ce93081016167dfd8e37e4e897bedf47d65#diff-3b74a927d8ddbc2c49178f751d0fd058

Note: NuGet.Build.Tasks.Pack.Library.csproj was renamed to NuGet.Build.Tasks.Pack.csproj

dominoFire replied:
This RuntimeIdentifier (RID) is here because at ilmerge.exe-time, System.Security.Cryptography.ProtectedData (S.S.C.PD) 4.3.0 only has assets for .NETFramework, not .NET Core. This RID helps ilmerge.exe to find the assembly. Upgrading to S.S.C.PD to 4.4.0+ removes the need of using this RID

Commit:
Revert "Removed VS feeds from NuGet.Config"

This reverts commit 811fcfc5c2ba1683c420101a9752aab76563eed3.

Commit:
Ref. assembly set to 1.0.0

Commit:
Removed VsToolsPath patch

nkolev92 commented on build/common.project.props:
```diff
@@ -19,8 +19,10 @@
     <NETCoreTestTargetFrameworks>netcoreapp3.1;netcoreapp5.0</NETCoreTestTargetFrameworks>
     <NetStandardVersion>netstandard2.0</NetStandardVersion>
     <TargetFrameworksExe>$(NETFXTargetFramework);$(NETCoreTargetFrameworks)</TargetFrameworksExe>
+    <TargetFrameworksExe Condition="'$(DotNetBuildFromSource)' == 'true'">$(NETCoreTargetFramework)</TargetFrameworksExe>
```
We use `IsBuildOnlyXPLATProjects` in a bunch of places.

dominoFire replied:
Made consistent.

nkolev92 commented on build/common.project.props:
```diff
@@ -19,8 +19,10 @@
+    <TargetFrameworksExe Condition="'$(DotNetBuildFromSource)' == 'true'">$(NETCoreTargetFramework)</TargetFrameworksExe>
     <TargetFrameworksLibrary  Condition=" '$(RequiresSigningXplatAPIs)' != 'true' ">$(NETFXTargetFramework);$(NetStandardVersion)</TargetFrameworksLibrary>
     <TargetFrameworksLibrary  Condition=" '$(RequiresSigningXplatAPIs)' == 'true' ">$(NETFXTargetFramework);$(NetStandardVersion);netcoreapp5.0</TargetFrameworksLibrary>
+    <TargetFrameworksLibrary Condition="'$(DotNetBuildFromSource)' == 'true'">$(NetStandardVersion)</TargetFrameworksLibrary>
```
Same here: IsBuildOnlyXPLATProjects.

dominoFire replied:
Done.

nkolev92 commented on build/common.props:
```diff
@@ -6,7 +6,7 @@
   <!-- When building on non-Windows machines, we need the reference assemblies to build full-framework assemblies. -->
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' AND '$(DotNetBuildFromSource)' == 'true'">
-     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0-alpha-5" PrivateAssets="All" />
+     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" PrivateAssets="All" />
```
I think this an oversight on my end. 
We should manage the version centrally.

dominoFire replied:
Ok, will move the version to packages.targets file

dominoFire replied:
Actually, I removed the reference assembly, because it seems it is not required, since source-build is building only netcore for 5.0; see https://github.com/NuGet/NuGet.Client/pull/3067#discussion_r326731073

nkolev92 commented on build/config.props:
```diff
@@ -27,6 +27,8 @@
     <NetCoreAssemblyBuildNumber Condition=" '$(NetCoreAssemblyBuildNumber)' == '' ">7</NetCoreAssemblyBuildNumber>
 
     <IsEscrowMode>false</IsEscrowMode>
+    <IsEscrowMode Condition="'$(DotNetBuildFromSource)' == 'true'">true</IsEscrowMode>
```
How does this affect source build?

dominoFire replied:
Naming conventions for the package, I believe. Is that true, @crummel  ?

nkolev92 replied:
Not really, check https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props, IsEscrow is only used for VS purposes.

dominoFire replied:
I've tested source-build without IsEscrowMode and it works. I will discard this patch

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj:
```diff
@@ -170,9 +170,6 @@
       <TfmSpecificPackageFile Include="$(OutputPath)\$(ILMergeSubpath)**\NuGet*.resources.dll">
         <PackagePath>CoreCLR/</PackagePath>
       </TfmSpecificPackageFile>
-      <TfmSpecificPackageFile Include="$(OutputPath)\$(ILMergeSubpath)**\*.dll" Condition="'$(IsBuildOnlyXPLATProjects)' == 'true'">
```
Why are we removing this? 

The one difference between source build and non source build sdks was that non source build ilmerged.

dominoFire replied:
It is adding additional NuGet assemblies that differs from the final sdk shipped by Microsoft. The difference was noted in this [comment](https://github.com/dotnet/source-build/issues/708#issuecomment-416818823). Tracking issue in source-build [here](https://github.com/dotnet/source-build/issues/739).

dominoFire replied:
In addition, there's a NU5118 warning in when NuGet is compiled via source-build (edited) when this patch is not applied:

```
/usr/local/share/dotnet/sdk/5.0.100-preview.4.20258.7/Sdks/NuGet.Build.Tasks.Pack/buildCrossTargeting/NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File '$REPO_ROOT/artifacts/NuGet.Build.Tasks.Pack/16.0/bin/Debug/netstandard2.0/NuGet.Build.Tasks.Pack.dll' is not added because the package already contains file 'CoreCLR/NuGet.Build.Tasks.Pack.dll' [$REPO_ROOT/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj]
```

nkolev92 replied:
Ok, my comment here is that at the time the current behavior was a deliberate decision.

nkolev92 commented on src/NuGet.Core/NuGet.Frameworks/NuGet.Frameworks.csproj:
```diff
@@ -4,7 +4,8 @@
 
   <PropertyGroup>
     <Description>NuGet's understanding of target frameworks.</Description>
-    <TargetFrameworks>$(TargetFrameworksLibrary);net40</TargetFrameworks>
```
Use `IsBuildOnlyXPLATProjects` or just make it consistent throughout. 


Honestly, I'd be in favor of removing net40 sooner rather than alter.

dominoFire replied:
Made consistent with `IsBuildOnlyXPLATProjects`

I will remove net40 in another PR https://github.com/NuGet/NuGet.Client/pull/3513

Commit:
Addressed feedback

- Substituted $DotnetBuildFromSource with $IsBuildOnlyXPLATProjects
- Moved netfx ref. assembly version to packages.targets

Commit:
Removed net40 form NuGet.Frameworks

Commit:
***NO_CI*** Removed IsEscrowMode patch

Commit:
Revert "Removed net40 form NuGet.Frameworks" ***NO_CI***

This reverts commit be0f0d8a2c7187c79296d1468c5d691297eed577.

Commit:
Revert "***NO_CI*** Removed IsEscrowMode patch" ***NO_CI**

This reverts commit 3fd448b5a955b67dcb64628b66a5601813f55b60.

Commit:
Moved properties

Commit:
Removed IsEscrowMode patch, again

Commit:
git status

Commit:
Removed ref. assembly from packages.targets

Commit:
***NO_CI*** Test removing RID patch in N.B.T.P

Commit:
Test without RuntimeIdentifier in N.B.T.P

Commit:
Diagnose ilmerge error

Commit:
Corrected newline for debugging ilmerge

Commit:
Upgrades System.Security.Cryptography.ProtectedData to remove RuntimeIdentifier in N.B.T.P

Commit:
Fixed dependencies in TestablePlugin

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj:
```diff
@@ -94,9 +95,19 @@
+        /keyfile:$(MS_PFX_PATH)</IlmergeCommand>
+      <IlmergeCommand>$(IlmergeCommand) ^
+        $(OutputPath)NuGet.Build.Tasks.Pack.dll ^
+        @(BuildArtifacts, ' ^&#x0A;')</IlmergeCommand>
```
Did `' '` cause problems? I'm, asking because this is unreadable.

dominoFire replied:
No. I added `' ^&#x0A;'` to add a new line in the command for the MSBuild Exec task to improve readability in the logs

nkolev92 replied:
What's the from and to in the logs?

dominoFire replied:
Original command --edited-- (From):

```
C:\repos\NuGet.Client\packages\ilmerge\3.0.21\tools\net452\ILMerge.exe /out:C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\ilmerge\NuGet.Build.Tasks.Pack.dll /targetplatform:v4,"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.7.2" /lib:C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\publish /internalize /xmldocs /log:C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\ilmergeIlMergeLog.txt /delaysign /keyfile:C:\repos\NuGet.Client\build\..\keys\35MSSharedLib1024.snk C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Build.Tasks.Pack.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\Newtonsoft.Json.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Commands.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Common.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Configuration.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Credentials.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.DependencyResolver.Core.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Frameworks.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.LibraryModel.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Packaging.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.ProjectModel.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Protocol.dll C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Versioning.dll
```

From MSBuild Structured Log Viewer (To):
```
CommandLineArguments = C:\repos\NuGet.Client\packages\ilmerge\3.0.21\tools\net452\ILMerge.exe ^
        /out:C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\ilmerge\NuGet.Build.Tasks.Pack.dll ^
        /targetplatform:v4,"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.7.2" ^
        /lib:C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\publish ^
        /internalize ^
        /xmldocs ^
        /log:C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\ilmergeIlMergeLog.txt ^
        /delaysign ^
        /keyfile:C:\repos\NuGet.Client\build\..\keys\35MSSharedLib1024.snk ^
        C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Build.Tasks.Pack.dll ^
        C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\Newtonsoft.Json.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Commands.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Common.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Configuration.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Credentials.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.DependencyResolver.Core.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Frameworks.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.LibraryModel.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Packaging.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.ProjectModel.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Protocol.dll ^
C:\repos\NuGet.Client\artifacts\NuGet.Build.Tasks.Pack\16.0\bin\Debug\net472\NuGet.Versioning.dll
```

nkolev92 replied:
I'm not seeing the impact of the weird char in this case. 

Anyway, don't want to block on this, my only point is because that's a very common MSBuild pattern `@(BuildArtifacts, ' ')`, not something that the NuGet team does, and this causes a pause for me reading it doesn't seem to be adding much.

dominoFire replied:
Ok, reverted back

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj:
```diff
@@ -111,9 +122,20 @@
+        /keyfile:$(MS_PFX_PATH)</IlmergeCommand>
+      <IlmergeCommand>$(IlmergeCommand) ^
+        $(OutputPath)$(CurrentCulture)NuGet.Build.Tasks.Pack.resources.dll ^
+        @(ResourceArtifacts, ' ^&#x0A;')</IlmergeCommand>
```
Did ' ' cause problems? I'm, asking because this is unreadable.

dominoFire replied:
Ibid. https://github.com/NuGet/NuGet.Client/pull/3511#discussion_r454570101

Commit:
Item metadata with spaces
